Argument and variables nullability

DifferentialOrange opened the pull request:
Reworked and improved version of https://github.com/tarantool/graphql/pull/59

Compare library with JS reference implementation in various nullability cases. This PR uses JS script to build various schemas and queries and execute them to generate luatest script that verifies that this library has the same behavior.

To generate script, run
```bash
cd ./tests/integration/codegen/fuzzing_nullability
npm install
node ./generate.js > ../../fuzzing_nullability_test.lua
```

This PR fixed several nullability behavior issues. There is some case that isn't yet fully covered by our library, refer to https://github.com/tarantool/graphql/issues/62 for updates.

Follow https://github.com/tarantool/graphql/issues/63 for more test cases.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -864,6 +885,8 @@ function accessor_general.new(opts, funcs)
                                      DEF_RESULTING_OBJECT_CNT_MAX
     local fetched_object_cnt_max = opts.fetched_object_cnt_max or
                                    DEF_FETCHED_OBJECT_CNT_MAX
+    -- todo: move this setting to `tgql.compile` after #59
```
Use TODO or XXX marks, they highlighted by vim.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -804,6 +809,22 @@ local function validate_funcs(funcs)
+        fetched_object_cnt = 0
+    }
+    qcontext.deadline_clock = clock.monotonic64() +
+                                settings.timeout_ms * 1000 * 1000
```
The indent is strange.

Totktonada commented on graphql/accessor_general.lua:
```diff
@@ -864,6 +885,8 @@ function accessor_general.new(opts, funcs)
     local fetched_object_cnt_max = opts.fetched_object_cnt_max or
                                    DEF_FETCHED_OBJECT_CNT_MAX
+    -- todo: move this setting to `tgql.compile` after #59
+    local timeout_ms = opts.timeout_ms or DEF_TIMEOUT_MS
```
I think we should check user-provided input.

Totktonada commented:
Added @SudoBobo as optional reviewer.

DifferentialOrange committed:
Propagate box.NULL result

Based on @no1seman PR [1].

1. https://github.com/tarantool/graphql/pull/59

DifferentialOrange committed:
Do not throw "no value provided" for box.NULL result

Based on @no1seman PR [1].

1. https://github.com/tarantool/graphql/pull/59

DifferentialOrange committed:
Fix variables coerce

Based on @no1seman PR [1].

1. https://github.com/tarantool/graphql/pull/59

DifferentialOrange committed:
Allow using defaults for Non-null variables

Specification does not explicitly forbid using defaults with Non-null
variables. The discussion [1] states that it is legal. Reference JS
implementation supports defaults for Non-null variables.

1. https://github.com/graphql/graphql-spec/issues/570

DifferentialOrange committed:
Forbid using null default for Non-nullable args

Specification [1] states that Nullable variable can be used for
NonNullable argument if it has Non-null default. See more in [2] issue.

1. http://spec.graphql.org/October2021/#sec-All-Variable-Usages-are-Allowed
2. https://github.com/graphql/graphql-js/issues/3715

DifferentialOrange committed:
Fix error message

Code uses NonNull wrapper to support case when Nullable variable has a
Non-null default value and is compatible with NonNullable argument. But
if it really incompatible, the user should receive an error with
original type in message.

DifferentialOrange committed:
Test argument and variable nullability

Compare library with JS reference implementation in various nullability
cases. This patch uses JS script to build various schemas and queries
and execute them to generate luatest script that verifies that this
library has the same behavior.

To generate script, run
```bash
cd ./tests/integration/codegen/fuzzing_nullability
npm install
node ./generate.js > ../../fuzzing_nullability_test.lua
```

Prior to this patch, several nullability behavior issues were fixed.
There is some case that isn't yet fully covered by our library, refer
to [1] for updates.

Follow [2] for more test cases.

Based on @no1seman PR [3].

1. https://github.com/tarantool/graphql/issues/62
2. https://github.com/tarantool/graphql/issues/63
3. https://github.com/tarantool/graphql/pull/59

DifferentialOrange commented:
Comment about https://github.com/tarantool/graphql/pull/64/commits/eaa9c3ccf78aff4ac791fe3c10ca446108442b2f : together with https://github.com/tarantool/graphql/pull/64/commits/033c1c9a31ce42ab29e1553541fb2ba2a095897b and https://github.com/tarantool/graphql/pull/64/commits/033c1c9a31ce42ab29e1553541fb2ba2a095897b it becomes useless, so actually we could remove it from the patchset.
